File: Sources/COCHelperCore/CoAPIError.swift
import Foundation

/// Errors surfaced by `CoAPIClient`, already mapped from HTTP status codes and
/// transport failures.
///
/// Associated values are sanitized: they never contain request URLs, headers,
/// raw bodies or tokens, so they are safe to log.
public enum CoAPIError: Error, Equatable, Sendable {
    case missingCredentials
    case unauthorized
    case accessDenied(reason: String)
    case notFound
    case rateLimited(retryAfterSeconds: Int?)
    case serverError(statusCode: Int)
    case timeout
    case network(underlying: String)
    case malformedResponse(detail: String)

    /// 结构化错误类别（Issue #252）：覆盖本类型全部 case **外加取消**
    /// （`EndpointRefresher` 单独捕获 `CancellationError` / `URLError(.cancelled)`，
    /// 不经由 `CoAPIError` 传递，故此处不包含 `.cancelled`）。
    ///
    /// 不含 associated value，可作为 `OfficialEndpointState.failureKind`
    /// 安全持久化——不泄露 `reason` / `detail` / `underlying` / `statusCode`
    /// 等脱敏细节，也不依赖中文文案稳定性。展示层仍由
    /// `lastErrorReason`（脱敏文本）承担，本属性只用于**分类判定**。
    public var kind: OfficialEndpointFailureKind {
        switch self {
        case .missingCredentials: return .missingCredentials
        case .unauthorized: return .unauthorized
        case .accessDenied: return .accessDenied
        case .notFound: return .notFound
        case .rateLimited: return .rateLimited
        case .serverError: return .serverError
        case .timeout: return .timeout
        case .network: return .network
        case .malformedResponse: return .malformedResponse
        }
    }
}

File: packages/domain/src/official/co-api-error.ts
import type { OfficialEndpointFailureKind } from './types';

export type CoAPIError =
  | { readonly kind: 'missingCredentials' }
  | { readonly kind: 'unauthorized' }
  | { readonly kind: 'accessDenied'; readonly reason: string }
  | { readonly kind: 'notFound' }
  | { readonly kind: 'rateLimited'; readonly retryAfterSeconds: number | undefined }
  | { readonly kind: 'serverError'; readonly statusCode: number }
  | { readonly kind: 'timeout' }
  | { readonly kind: 'network'; readonly underlying: string }
  | { readonly kind: 'malformedResponse'; readonly detail: string };

export class CoAPIRequestCancelledError extends Error {
  override readonly name = 'CoAPIRequestCancelledError';
}

export function isCoAPIRequestCancelled(error: unknown): boolean {
  return error instanceof CoAPIRequestCancelledError || error instanceof DOMException && error.name === 'AbortError';
}

export function coAPIErrorKind(error: CoAPIError): OfficialEndpointFailureKind {
  switch (error.kind) {
    case 'missingCredentials':
      return 'missingCredentials';
    case 'unauthorized':
      return 'unauthorized';
    case 'accessDenied':
      return 'accessDenied';
    case 'notFound':
      return 'notFound';
    case 'rateLimited':
      return 'rateLimited';
    case 'serverError':
      return 'serverError';
    case 'timeout':
      return 'timeout';
    case 'network':
      return 'network';
    case 'malformedResponse':
      return 'malformedResponse';
  }
}

export function coAPIErrorHttpStatus(error: CoAPIError): number | undefined {
  switch (error.kind) {
    case 'unauthorized':
      return 401;
    case 'accessDenied':
      return 403;
    case 'notFound':
      return 404;
    case 'rateLimited':
      return 429;
    case 'serverError':
      return error.statusCode;
    default:
      return undefined;
  }
}

export function coAPIErrorUserFacingReason(error: CoAPIError): string {
  switch (error.kind) {
    case 'missingCredentials':
      return '未配置 API token';
    case 'unauthorized':
      return '认证失败（401）';
    case 'accessDenied':
      return `访问被拒绝：${error.reason}`;
    case 'notFound':
      return '未找到对应的玩家、部落或部落对战（404）';
    case 'rateLimited':
      return '请求被限流（429），请稍后再试';
    case 'serverError':
      return `服务器错误（${error.statusCode}）`;
    case 'timeout':
      return '请求超时';
    case 'network':
      return `网络错误（${error.underlying}）`;
    case 'malformedResponse':
      return `响应解析失败（${error.detail}）`;
  }
}

export function coAPIErrorsEqual(left: CoAPIError, right: CoAPIError): boolean {
  if (left.kind !== right.kind) {
    return false;
  }
  switch (left.kind) {
    case 'accessDenied':
      return right.kind === 'accessDenied' && left.reason === right.reason;
    case 'rateLimited':
      return right.kind === 'rateLimited' && left.retryAfterSeconds === right.retryAfterSeconds;
    case 'serverError':
      return right.kind === 'serverError' && left.statusCode === right.statusCode;
    case 'network':
      return right.kind === 'network' && left.underlying === right.underlying;
    case 'malformedResponse':
      return right.kind === 'malformedResponse' && left.detail === right.detail;
    default:
      return true;
  }
}
